specs/execution-summary.md

```diff
@@ -0,0 +1,178 @@
+# Execution summary
+
+Execution summary gives users a quick overview of the requests intercepted by the proxy and metrics about the number and types of requests messages logged by the proxy. This feature retrieves information about requests from [recorded request logs](./recording-requests.md).
+
+## History
+
+| Version | Date | Comments | Author |
+| ------- | ---- | -------- | ------ |
+| 1.0 | 2023-02-06 | Initial specifications | @waldekmastykarz |
+
+## Implementation
+
+### Information source
+
+Execution summary is a proxy plugin subscribed to the `RecordingStopped` event emitted by the request logs recorder. This event contains information about the request log messages captured during the recording session. The execution summary plugin uses this information to generate a summary of the requests captured during the recording session. The summary uses Markdown syntax to format the output.
+
+### Configuration
+
+Setting|Option|Description
+-------|-----------------|-----------
+`filePath`|`--summary-file-path`|Path to the file where the summary should be saved. If not specified, the summary will be printed to the console. Path can be absolute or relative to the current working directory. Default: empty.
+`groupBy`|`--summary-group-by`|Specifies how the information should be grouped in the summary. Available options: `url` (default), `messageType`.
+
+### Type of information
+
+Execution summary contains the following information:
+
+- Summary title (fixed)
+- Date and time of the recording session
+- Groups as specified by the `groupBy` configuration option:
+  - if grouped by URL:
+    - List of requests captured during the recording session sorted alphabetically. For each request:
+      - Request method
+      - Request URL
+      - List of unique request log messages for this request issued by the proxy, grouped and sorted by severity. Excluding the `InterceptedRequest` and `PassedThrough` severity.
```

Comment:
Can we also have the number of instances for this log message? I feel it would be relevant to know that we had 22 WARNINGS on the /me/directReports for instance.

Reply:
Good idea. Shall we add the count after each message, eg.

```md
### GET https://graph.microsoft.com/v1.0/me

#### Warning

- To improve performance of your application, use the $select parameter. ore info at https://learn.microsoft.com/graph/query-parameters#select-parameter (22)
```

Reply:
```suggestion
      - (22) To improve performance of your application, use the $select parameter. ore info at https://learn.microsoft.com/graph/query-parameters#select-parameter (22)
```

Reply:
Would you display the number of occurrences both before and after? 😕

Reply:
Bad copy/paste haha! Just before!

Reply:
When grouping per URL it might make sense to provide CSV exports as well. Had a internal discussion with some testers who considered testing the application for extended periods op time (so lets say you have a complex process that consists of 15 steps you might want to know what happens in each step when 10% of our calls fail). Using the proxy you facilitate that scenario but with the report a CSV with all failed requests that contains `TimeStamp, Request Method, Request URL, Error Message` devs can map errors to what might be logged in App Insights to get even more details and do better analysis. I assume it should be a specific configuration option and the CSV should live next to the markdown output.

Reply:
Would the CSV be just another output form next to console or would it be a summary? In other words: would you expect to see in the CSV all occurrences of the same error or only the unique occurrences?